### PR TITLE
PRIME-1733 Health Authority Updates

### PR DESCRIPTION
**Pull request by mtpultz**
**Health Authority Testing**
1. Locally pull the develop branch and without running this branches migrations create a number of community sites using the old schema
2. Switch branches to this branch and run the migrations, and check that everything in the DB is as expected for community sites
3. Create more community sites and make sure the workflow works as expected
4. Create health authority sites and make sure the workflow functions as expected (see below for functionality and states)
5. Verify that admin functionality around sites works as expected

**Expected Functionality**
* Able to create and update a site's information
* Able to submit and resubmit a site
* Able to reach expected views based on site state
* Unable to reach specific views based on site state

**Health Authority Site States**
These are the different kinds of states based on status and the combination of submitted and approved date:

1.  Editable, Incomplete (has not reached overview), and Not Submitted
2. Editable, Complete (reached overview at least once), and Not Submitted
3. In Review
4. Editable and Submitted which occurs when placed back into Editable for updates by an admin (Requires status to be added to the DB manually)
5. Locked (Requires status to be added to the DB manually)
6. Approved (Requires status and approved date to be added to the DB manually

### CODE REVIEW

File: prime-angular-frontend/src/app/modules/gis-enrolment/shared/services/gis-enrolment.service.ts [hunk 1]
```diff
@@ -37,7 +37,7 @@ export class GisEnrolmentService implements IEnrolmentService<GisEnrolment> {
   }
 
   public get enrolment$(): Observable<GisEnrolment> {
-    // Allow subscriptions, but make immutable
+    // Allow subscriptions, but make the subject immutable
```
Comment (AMiKaH) [inline prime-angular-frontend/src/app/modules/gis-enrolment/shared/services/gis-enrolment.service.ts:40]: subject or object?

File: prime-angular-frontend/src/app/modules/health-auth-site-reg/shared/guards/health-authority-site-guard.service.ts [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+  ) {
+    super(authService, logger);
+  }
+
+  protected checkAccess(routePath: string, params?: Params): Observable<boolean> | Promise<boolean> {
+    return this.healthAuthorityResource.getHealthAuthoritySiteById(params.haid, params.sid)
+      .pipe(
+        map((healthAuthoritySite: HealthAuthoritySite) =>
+          // Store the site for access throughout registration, which
+          // will allows be the most up-to-date site information
```
Comment (AMiKaH) [inline prime-angular-frontend/src/app/modules/health-auth-site-reg/shared/guards/health-authority-site-guard.service.ts:41]: Sentence missing words? "will allows be"

File: prime-angular-frontend/src/app/modules/health-auth-site-reg/shared/services/authorized-user.service.ts [hunk 3]
```diff
@@ -38,7 +37,7 @@ export class AuthorizedUserService implements IAuthorizedUserService {
   }
 
   public get authorizedUser$(): Observable<AuthorizedUser | null> {
-    // Allow subscriptions, but make immutable
+    // Allow subscriptions, but make the subject immutable
```
Comment (AMiKaH) [inline prime-angular-frontend/src/app/modules/health-auth-site-reg/shared/services/authorized-user.service.ts:40]: Same as above, subject or object?

File: prime-angular-frontend/src/app/modules/health-auth-site-reg/shared/services/health-authority-site.service.ts [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+  public set site(healthAuthoritySite: HealthAuthoritySite) {
+    this._site.next(healthAuthoritySite);
+  }
+
+  public get site(): HealthAuthoritySite | null {
+    return this._site.value ?? null;
+  }
+
+  public get site$(): Observable<HealthAuthoritySite | null> {
+    // Allow subscriptions, but make the subject immutable
```
Comment (AMiKaH) [inline prime-angular-frontend/src/app/modules/health-auth-site-reg/shared/services/health-authority-site.service.ts:32]: Same, subject vs object

File: prime-dotnet-webapi/Services/interfaces/IHealthAuthorityService.cs [hunk 5]
```diff
@@ -14,9 +15,14 @@ public interface IHealthAuthorityService
         Task<IEnumerable<HealthAuthorityListViewModel>> GetHealthAuthoritiesAsync();
         Task<HealthAuthorityViewModel> GetHealthAuthorityAsync(int healthAuthorityId);
         Task<IEnumerable<AuthorizedUserViewModel>> GetAuthorizedUsersAsync(int healthAuthorityId);
+        Task<bool> AuthorizedUserExistsOnHealthAuthorityAsync(int healthAuthorityId, int authorizedUserId);
         Task UpdateCareTypesAsync(int healthAuthorityId, IEnumerable<string> careTypes);
         Task UpdateContactsAsync<T>(int healthAuthorityId, IEnumerable<ContactViewModel> contacts) where T : HealthAuthorityContact, new();
         Task UpdatePrivacyOfficeAsync(int healthAuthorityId, PrivacyOfficeViewModel privacyOffice);
         Task UpdateVendorsAsync(int healthAuthorityId, IEnumerable<int> vendorCodes);
+
```
Comment (AMiKaH) [inline prime-dotnet-webapi/Services/interfaces/IHealthAuthorityService.cs:23]: Extra space

File: prime-dotnet-webapi/ViewModels/Sites/RemoteUserViewModel.cs [hunk 6]
```diff
@@ -0,0 +1,14 @@
...
+
+namespace Prime.ViewModels.Sites
+{
+    public class RemoteUserViewModel
+    {
+        public int Id { get; set; }
+        public string FirstName { get; set; }
+        public string LastName { get; set; }
+        public string Email { get; set; }
+
```
Comment (AMiKaH) [inline prime-dotnet-webapi/ViewModels/Sites/RemoteUserViewModel.cs:11]: Extra space

File: prime-dotnet-webapi/ViewModels/Sites/_CommonSiteMappingProfile.cs [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+    {
+        public CommonSiteMappingProfile()
+        {
+            CreateMap<BusinessDay, BusinessDayViewModel>()
+                .ReverseMap();
+            CreateMap<RemoteUser, RemoteUserViewModel>()
+                .ReverseMap();
+            CreateMap<RemoteUserCertification, RemoteUserCertificationViewModel>()
+                .ReverseMap();
+
```
Comment (AMiKaH) [inline prime-dotnet-webapi/ViewModels/Sites/_CommonSiteMappingProfile.cs:19]: Extra space

File: prime-angular-frontend/src/app/modules/health-auth-site-reg/shared/services/authorized-user.service.ts (hunk 3 above)
Comment (mtpultz) [reply prime-angular-frontend/src/app/modules/health-auth-site-reg/shared/services/authorized-user.service.ts:40]: No, it's a BehaviourSubject being made immutable

File: prime-angular-frontend/src/app/modules/gis-enrolment/shared/services/gis-enrolment.service.ts (hunk 1 above)
Comment (mtpultz) [reply prime-angular-frontend/src/app/modules/gis-enrolment/shared/services/gis-enrolment.service.ts:40]: No, it's a BehaviourSubject being made immutable

**Review by AMiKaH — APPROVED** [review]

**Review by james-hollinger — APPROVED** [review]
THIS PR IS #! 
⁽ᶦⁿ ᶜʰᵃⁿᵍᵉᵈ ᶠᶦˡᵉˢ ᵃˢ ᵒᶠ ²⁰²¹⁻¹¹⁻⁰¹⁾

**Pull request merged**